### PR TITLE
[X86][GlobalISel] Fix referencing nonexistent operand in G_ICMP

### LINKED ISSUE #86203: [X86] GlobalIsel removed type info when legalizing ICMP, causing out of bound access on debug build
## Reproduction 

[Out of bound on debug build](https://godbolt.org/z/zTPP5jbTG) 
[Unable to legalize on release build](https://godbolt.org/z/hrnsaofYY)

The bug can be reproduced on LLVM 16, 17, and upstream.

@phoebewang  Any hints on how should we fix this?

### PR DESCRIPTION
**Pull request by e-kud**
Fixes #86203

### CODE REVIEW

File: llvm/test/CodeGen/X86/GlobalISel/legalize-icmp-vec.mir [hunk 1]
```diff
@@ -0,0 +1,24 @@
+# RUN: llc -mtriple=x86_64-linux-gnu -run-pass=legalizer -global-isel-abort=2 -pass-remarks-missed='gisel*' %s -o - 2>%t | FileCheck %s
+# RUN: FileCheck -check-prefix=ILLEGAL %s < %t
+
+# ILLEGAL: remark: <unknown>:0:0: unable to legalize instruction: %2:_(<4 x s1>) = G_ICMP intpred(sle), %0:_(<4 x s64>), %1:_ (in function: test_icmp_v4i64)
```
Comment (phoebewang) [inline llvm/test/CodeGen/X86/GlobalISel/legalize-icmp-vec.mir:4]: Is this patch only to fix the first crash reported in #86203?

**Review by arsenm — APPROVED** [review]

File: llvm/test/CodeGen/X86/GlobalISel/legalize-icmp-vec.mir (hunk 1 above)
Comment (RKSimon) [reply llvm/test/CodeGen/X86/GlobalISel/legalize-icmp-vec.mir:4]: Please can you add a reference to PR86203?

Comment (e-kud) [reply llvm/test/CodeGen/X86/GlobalISel/legalize-icmp-vec.mir:4]: @phoebewang it's not a crash, actually. The crash happens if we use only `-global-isel`. This option forces llc to use only GlobalISel. The correct way is to use `-fglobal-isel` with clang or `-global-isel -global-isel-abort=2` with llc. It allows to fallback to SelectionDAG if something is unsupported in GlobalISel. Without this fix we hit the assert and are not able to fallback to SelectionDAG.

@RKSimon done.

**Review by RKSimon — APPROVED** [review]
LGTM

**Comment by AtariDreams** [comment]
/cherry-pick https://github.com/llvm/llvm-project/commit/fb394562a3cf3522ac17a1436234ef5b926636fb

**Comment by topperc** [comment]
@AtariDreams GlobalISel is still very much in development for X86. Why do you think this needs to be backported?

**Pull request merged**